### ISSUE TITLE
Figure out how to adapt to Bytes' introduction in GHC 8.10

### ISSUE THREAD

**Issue by RyanGlScott** [mgsloan/th-orphans#27]
GHC 8.10's version of `template-haskell` will feature a new `Bytes` data type (introduced in [this commit](https://gitlab.haskell.org/ghc/ghc/commit/224a6b864c6aa0d851fcbf79469e5702b1116dbc)):

```haskell
data Bytes = Bytes
   { bytesPtr    :: ForeignPtr Word8 -- ^ Pointer to the data
   , bytesOffset :: Word             -- ^ Offset from the pointer
   , bytesSize   :: Word             -- ^ Number of bytes
   }

data Lit
  = ...
  | BytesPrimL Bytes     -- ^ Some raw bytes, type Addr#:`
```

It turns out that this change breaks `th-orphans`:

```
[1 of 1] Compiling Language.Haskell.TH.Instances ( src/Language/Haskell/TH/Instances.hs, /home/rgscott/Documents/Hacking/Haskell/cabal-sandbox/dist-newstyle/build/x86_64-linux/ghc-8.9.20190309/th-orphans-0.13.6/build/Language/Haskell/TH/Instances.o )

src/Language/Haskell/TH/Instances.hs:520:3: error:
    • Couldn't match a lifted type with an unlifted type
      When matching types
        t0 :: *
        ghc-prim-0.5.3:GHC.Prim.Weak# () :: TYPE
                                              'ghc-prim-0.5.3:GHC.Types.UnliftedRep
    • In the first argument of ‘lift’, namely ‘x0_ackI’
      In the second argument of ‘appE’, namely ‘(lift x0_ackI)’
      In the expression:
        (appE
           (conE
              ((Name (mkOccName "CFinalizers"))
                 (((NameG DataName) (mkPkgName "base"))
                    (mkModName "GHC.ForeignPtr")))))
          (lift x0_ackI)
    |
520 | $(reifyManyWithoutInstances ''Lift [''Info, ''Loc] (const True) >>=
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

src/Language/Haskell/TH/Instances.hs:520:3: error:
    • Couldn't match a lifted type with an unlifted type
      When matching types
        t1 :: *
        ghc-prim-0.5.3:GHC.Prim.MutVar# s a :: TYPE
                                                 'ghc-prim-0.5.3:GHC.Types.UnliftedRep
    • In the first argument of ‘lift’, namely ‘x0_ackG’
      In the second argument of ‘appE’, namely ‘(lift x0_ackG)’
      In the expression:
        (appE
           (conE
              ((Name (mkOccName "STRef"))
                 (((NameG DataName) (mkPkgName "base")) (mkModName "GHC.STRef")))))
          (lift x0_ackG)
    • Relevant bindings include
        x0_ackG :: ghc-prim-0.5.3:GHC.Prim.MutVar# s a
          (bound at src/Language/Haskell/TH/Instances.hs:520:3)
        e_ackF :: GHC.STRef.STRef s a
          (bound at src/Language/Haskell/TH/Instances.hs:520:3)
        lift :: GHC.STRef.STRef s a -> Q Exp
          (bound at src/Language/Haskell/TH/Instances.hs:520:3)
    |
520 | $(reifyManyWithoutInstances ''Lift [''Info, ''Loc] (const True) >>=
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

src/Language/Haskell/TH/Instances.hs:520:3: error:
    • Couldn't match a lifted type with an unlifted type
      When matching types
        t2 :: *
        ghc-prim-0.5.3:GHC.Prim.MutableByteArray#
          ghc-prim-0.5.3:GHC.Prim.RealWorld :: TYPE
                                                 'ghc-prim-0.5.3:GHC.Types.UnliftedRep
    • In the first argument of ‘lift’, namely ‘x0_ackA’
      In the second argument of ‘appE’, namely ‘(lift x0_ackA)’
      In the first argument of ‘appE’, namely
        ‘((appE
             (conE
                ((Name (mkOccName "MallocPtr"))
                   (((NameG DataName) (mkPkgName "base"))
                      (mkModName "GHC.ForeignPtr")))))
            (lift x0_ackA))’
    |
520 | $(reifyManyWithoutInstances ''Lift [''Info, ''Loc] (const True) >>=
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

This because `th-orphans` tries to automatically generate a `Lift` instance for `Bytes` through Template Haskell. Alas, `ForeignPtr` can't be given a `Lift` instance, so this is troublesome.

A quick-and-dirty fix would be to simply define a partial `Lift Bytes` instance:

```diff
diff --git a/src/Language/Haskell/TH/Instances.hs b/src/Language/Haskell/TH/Instances.hs
index bd4d85b..11ce01a 100644
--- a/src/Language/Haskell/TH/Instances.hs
+++ b/src/Language/Haskell/TH/Instances.hs
@@ -517,5 +517,10 @@ deriving instance Typeable Ppr
 deriving instance Typeable Quasi
 #endif
 
+#if __GLASGOW_HASKELL__ >= 809
+instance Lift Bytes where
+  lift (Bytes _fp off sz) = [| Bytes (error "Cannot lift ForeignPtr") off sz |]
+#endif
+
 $(reifyManyWithoutInstances ''Lift [''Info, ''Loc] (const True) >>=
   deriveLiftMany)
```

I'm not sure if there's a more principled solution.

**Comment by mgsloan** [comment]
Thanks for bringing this up!  I think something like this ought to do the trick, but I haven't been able to try compiling it yet:

```haskell
#if MIN_VERSION_template_haskell(2,15,0)
import GHC.ForeignPtr (ForeignPtr(ForeignPtr), Finalizers(NoFinalizers))
import System.IO.Unsafe (unsafePerformIO)
#endif

#if MIN_VERSION_template_haskell(2,15,0)
instance Lift Bytes where
  lift bytes =
    [| Bytes
      { bytesPtr = ForeignPtr $(litE (BytesPrimL bytes)) (unsafePerformIO (newIORef NoFinalizers))
      , bytesOffset = 0
      , bytesSize = bytesSize bytes
      } |]
#endif
```

I haven't been able to test it yet because I can't find the GHC change described - the link to the commit doesn't resolve.  I'm also not finding `BytesPrimL` on gitlab or my local ghc checkout.  Might be related to the ongoing trac migration.

**Comment by RyanGlScott** [comment]
> I haven't tested this yet - seems like https://gitlab.haskell.org/ghc/ghc/merge_requests/141/diffs#e0300ef07066368947e579d6342b69d6d88f881c_0_18 isn't merged yet? The merge commit you linked 404s for me.

I think this was due to the Trac-to-GitLab migration. The site is back up now.

**Comment by mgsloan** [comment]
@RyanGlScott It seems to work! https://github.com/mgsloan/th-orphans/commit/de4613cd45cffb4b7a4c54df5f6b74952279e5e9

What do you think, perhaps I should cut a release with this?

**Comment by RyanGlScott** [comment]
Nice! This just goes to show that `unsafePerformIO` is the both the cause and solution to all of life's problems :)

> What do you think, perhaps I should cut a release with this?

I don't think we have to be in a big hurry to release this, since `Bytes` won't debut until two full GHC releases from now. Plus, there is a chance that `Bytes`' definition could change from now until the 8.10 release.

**Comment by mgsloan** [comment]
@RyanGlScott Thanks for reviewing!  I've pushed some more commits

**Comment by mgsloan** [comment]
Closing - this was included in version `0.13.7`
